File: app/resources.py
"""Non-blocking resource telemetry for the local workstation."""
import os
import time

import psutil


class ResourceMonitor:
    def __init__(self):
        self.process = psutil.Process()
        self.at = 0.0
        self.current = {}
        psutil.cpu_percent(interval=None)

    def snapshot(self):
        if time.monotonic() - self.at >= 2:
            memory = psutil.virtual_memory()
            self.current = {"logical_cpus": os.cpu_count() or 1,
                            "cpu_percent": psutil.cpu_percent(interval=None),
                            "memory_total_gib": round(memory.total / 2**30, 1),
                            "memory_used_gib": round(memory.used / 2**30, 1),
                            "memory_percent": memory.percent,
                            "server_memory_mib": round(self.process.memory_info().rss / 2**20, 1)}
            self.at = time.monotonic()
        return dict(self.current)

File: src/resources.js
import os from 'node:os';
export class ResourceMonitor {
  constructor(){this.previous=null;this.at=0;this.current={};}
  snapshot(){
    if(Date.now()-this.at<2000)return this.current;
    const cpus=os.cpus(),total=cpus.reduce((n,c)=>n+Object.values(c.times).reduce((a,b)=>a+b,0),0),idle=cpus.reduce((n,c)=>n+c.times.idle,0);
    const percent=this.previous&&total>this.previous.total?100*(1-(idle-this.previous.idle)/(total-this.previous.total)):0;
    this.previous={total,idle};this.at=Date.now();
    this.current={logical_cpus:cpus.length,cpu_percent:Math.max(0,Math.min(100,percent)),memory_total_gib:os.totalmem()/2**30,memory_used_gib:(os.totalmem()-os.freemem())/2**30,memory_percent:100*(1-os.freemem()/os.totalmem()),server_memory_mib:process.memoryUsage().rss/2**20};return this.current;
  }
}
